### DISCUSSION TITLE
チャンネルの連合処理を実装するに当たって考えるべきことを整理するためのスレッド

### DISCUSSION THREAD

**Discussion by KisaragiEffective** [misskey-dev/misskey#14049] [General]
## 整理の目的
以下の2つをメインに考えたい

* リモートのチャンネルに対するモデレーションとして何がほしいか
* リモートのチャンネルへの投稿が連合された場合、どのような情報がほしいか

## このissueを書いた時点での自分の見解

* チャンネルの連合処理 → チャンネル管理者による選択式が望ましそう (related: #10710)
  * デフォルト: 今までどおり連合させず、`localOnly = true` を強制する
  * 想定
     1. 「連合ありノートを許可する」スイッチを新設
     2. チャンネル管理者がオプトインする
     3. 投稿者は連合ありが許可されている場合連合ありノートをできるようになる
* チャンネル自体の情報の連合 → あってほしそう
  * 想定ケース: isSensitiveが立っているチャンネルの場合、LTLに露出させたくない (cf. c2383fac16306f51d045f9cec4f56430633291fa)
* 同時に追加したほうが良さそうな機能
  * backend: 流入してくるチャンネルへのモデレーション機能としてサーバー管理と同等の機能を追加すると良さそう？
    * チャンネルをサイレンス → チャンネルを強制的にホーム送りにする
    * チャンネルの連合をカット → そのチャンネルで行われたノートをすべてドロップ
  * frontend: チャンネルからのノートは`${チャンネル名} from ${インスタンス名}`のように、他のインスタンスから流れてきたことがわかるラベルをつける (デザイン案1)

## このissueを閉じるべきとき
* ~~ソフトウェアとしてチャンネルの連合処理を実装しないという判断がされた場合~~ (ref. https://github.com/misskey-dev/misskey/issues/10637#issuecomment-1550620956) → 今も変わりはない？
* チャンネルの連合処理が一応実装された場合
  * 実装後の新規機能追加要望などはissueのスレッドが散らかるので個別にissueが立てられるべき
  
## 添付ファイル
デザイン案1: チャンネルに投稿されて連合されてきたノートがフロントエンド上でどのように表示されるべきか

![image](https://github.com/misskey-dev/misskey/assets/48310258/1e9fba71-8827-4d85-ba42-403f74836773)

cc #8475 #11400

### Purpose

適当に実装するとやばそうなので実装前に意見を募集したい (実装できるとは言っていない)

### Do you want to implement this feature yourself?

- [ ] Yes, I will implement this by myself and send a pull request

**Comment by KisaragiEffective** [comment]
prior discussion: https://misskey.niri.la/notes/9uq9bdz57h and its subtree (非公式)

**Comment by anatawa12** [comment]
需要として想定されそうなものという意味ではリモートのチャンネルをフォローしたいという需要はありそう。

すぐにやらなくてもこれを想定した実装にしたい気がする

**Reply by KisaragiEffective** [reply]
> リモートのチャンネルをフォローしたい

実は #11400 で既出

**Comment by KisaragiEffective** [comment]
今までのAPIとの一貫性を考えると、リモートユーザーのように「リモートチャンネル」という概念を生やす必要がありそう

**Reply by KisaragiEffective** [reply]
リモートチャンネルを生やすとUpdateコマンドを送信するべきだが、そのペイロードはカスタムになってしまうのでは…？

**Reply by Sayamame-beans** [reply]
Misskeyのファイルの連合に近い話…?

**Reply by KisaragiEffective** [reply]
アレはattachmentsでなんとかごまかせるけどこれはそうもいかない

**Comment by syuilo** [comment]
チャンネルの管理者のアカウントがあるサーバー自体が閉鎖した際に残されたチャンネルはどうなるのかしら

**Reply by KisaragiEffective** [reply]
Deleteコマンドを連合先に配る？

**Reply by anatawa12** [reply]
リモートユーザと同じように何もしないと何も消えない形も一つの方法だとは思う

**Reply by syuilo** [reply]
その場合そのチャンネル内の投稿がサーバーによってばらばらになりそう？

**Reply by KisaragiEffective** [reply]
Deleteコマンドを連合先に配るだと負荷が高そうなので、手元のサーバーから配送して410が出たときにアーカイブされたフラグを立てるのほうがいいかも

**Reply by syuilo** [reply]
Aサーバーで作られたチャンネルXがBサーバーとCサーバーにも連合されていて、Ａが消えたときにＢがＸに投稿してもＣはそれを知る術が無さそう
ＢもＸがＣに連合されていることは知らない

**Reply by KisaragiEffective** [reply]
XにA以外から書き込めるような拡張をするのは考えていなかった
(面白そうだけど多分割に合わない)

**Reply by anatawa12** [reply]
チャンネルの主のサーバーに連合できなければ(チャンネル主のサーバーがsuspendされてれば)そのチャンネルに投稿を禁止するべきではありそう

> Deleteコマンドを連合先に配るだと負荷が高そうなので、手元のサーバーから配送して410が出たときにアーカイブされたフラグを立てるのほうがいいかも

にかぶるけど

**Reply by anatawa12** [reply]
> XにA以外から書き込めるような拡張をするのは考えていなかった
(面白そうだけど多分割に合わない)

リプライの処理で多分必要になりそう

**Reply by Sayamame-beans** [reply]
誰もフォローしていないリモートのチャンネルへは投稿不可にする…?

チャンネルをフォローすることでチャンネルの情報を得るという前提に立つと、B,Cの誰かがXをフォローしていればチャンネルがアーカイブされたことを知る術がある(投稿可否が分かる)ので…
(場合によってはproxyアカウントにフォローさせられるかも)

しかしこれだと、リプライしたくなった時に困りそうですね…

**Reply by Sayamame-beans** [reply]
#14030 と同じ感じで、"届かない可能性がある"を示すと多少は良さそうですかね?

**Comment by syuilo** [comment]
Aサーバーで作られたチャンネルXがBサーバーとCサーバーにも連合されていて、ＸにＢのユーザーが投稿したときにＡはＣに対してアクティビティをリレーするような形になる？

**Reply by KisaragiEffective** [reply]
多分そう (#11400 で軽量なリレーじゃない？という指摘が入っている)

**Reply by Sayamame-beans** [reply]
"チャンネルをフォローする"ことでそのチャンネルの投稿を拾う形を取ると、そのようになりそうです
(フォローしていない場合は、リモートの投稿同様、(ユーザーの投稿として拾うことで)観測したものだけ…)

**Comment by Sayamame-beans** [comment]
チャンネルのアーカイブ仕様メモ
- 当該チャンネルは検索に含まれないようになる
- 直接アクセスは可能
- タイムラインは(長期間経つと)そのうち見られなくなる
- チャンネル内への新規投稿(リプライ含む)が不可に
- チャンネル内のノートは消えない
- リアクションは可能
- チャンネル外へのRNや引用は可能
- チャンネル外からチャンネル内のノートを引用したものに対しては、アーカイブ後もリプライ等可能

**Reply by Sayamame-beans** [reply]
- アーカイブ化は(API経由で)解除出来る
(この挙動はお知らせのアーカイブ化と同様)

**Comment by KisaragiEffective** [comment]
cf. https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md

**Comment by KisaragiEffective** [comment]
https://misskey.niri.la/notes/9vqs1vadvu
> 流す情報の必要性についてスコープや見通しが立っていないからめちゃめちゃ適当だけど、とりあえずチャンネル情報を向こうに流したいよねってことは一致してそうなので、ひとまずFEP-1b12をシカトして`type: "Note"`なIObjectに`_misskey_channel: MiChannel`を増築することになりそう

**Reply by tai-cha** [reply]
これに関してはFEP-1b12をシカトしなくてもFEP-1b12の拡張でもできそうなのでシカトすると決めるには早いかもしれない？

_Ordinaly posted in_: https://mi.taichan.site/notes/9vqsczqqa4

**Comment by tesaguri** [comment]
チャンネルのノートを連合なしにするという現在の仕様に至った理由がチャンネルのノートの文脈がチャンネル内に閉じているであろうからというものであったこと（#6669）を踏まえると、チャンネルを連合させるに当たってチャンネルのノートを照会したときにそのノートがどのチャンネルに宛てたものであるかが分かるようになっているのが望ましいと思います。これについてはFEP-1b12で[規定されている](https://w3id.org/fep/1b12#audience-property)ように、`Note`の`audience`プロパティでチャンネルを指定することで素直に実現できるかと思います。

これをさらに飛躍させたアイデアとして、完全なチャンネルの連合機能を実装する前にこの文脈指定の機能のみを実現できるのではないかといったことを考えています。具体的には、`inbox`を持たない（つまり、リモートからのフォローを受け付けない）`Group`としてチャンネルのメタデータのみを連合して、`Note`の`audience`プロパティでその`Group`を指すといった感じです。将来的に完全なチャンネルの連合機能が実現した場合はこの`Group`に`inbox`を生やすことで"不完全な"チャンネルと区別できるでしょう。飛び道具的な対応という感は否めませんが、もしこれが実現すればチャンネルのノートを連合なしにする現在の制限を早期に緩和できるのではないかと思います（#10710）。

**Comment by tai-cha** [comment]
チャンネルに関しては全体のモデレーション責任をチャンネルを保有するインスタンスが行うのかあくまでノートを発行するインスタンスが行うのか、その両方なのかに応じて（チャンネルからの）ノートの削除が可能なインスタンスを考えるなどの観点も必要そうな気がしてきました

**Comment by KisaragiEffective** [comment]
https://github.com/misskey-dev/misskey-hub-next/pull/289
ひとまずたたき台
